Standards/scs-0211-v2-kaas-default-storage-class.md

```diff
@@ -0,0 +1,59 @@
+---
+title: SCS KaaS default storage class
+type: Standard
+status: Draft
+replaces: scs-0211-v1-kaas-default-storage-class.md 
+track: KaaS
+description: |
+  The SCS-0211 standard ensures that a default StorageClass with specific characteristics is available to KaaS users.
+---
+
+## Introduction
+
+This is the standard v2 for SCS Release 7.
+
+Cluster consumers can request persistent storage via [`PersistentVolumeClaims`][k8s-pvc], which is provisioned
+automatically by cloud-provided automation.
+Storage requirements may vary across use cases, so there is the concept of storage classes (`StorageClass`).
+Storage classes define some set of storage properties and consumers can choose one of these depending on the use case.
+
+## Motivation
+
+A lot of third-party software, such as Helm charts, assume that a default storage class is configured.
+Thus, for an out-of-the-box working experience, a SCS compliant Kubernetes cluster should come
+preconfigured with a sensible default storage class providing persistent storage.
+
+## Decision
+
+A freshly provisioned Kubernetes cluster MUST have a default storage class, i.e., a `StorageClass`
+object that is annotated with `storageclass.kubernetes.io/is-default-class=true` as described in the
+[Kubernetes documentation][k8s-default-sc].
+The name of this storage class is not standardized.
+
+The persistent volumes (PV) provisioned by the provided default storage class MUST fulfill all
+of the following properties:
+
+- MUST support the `ReadWriteOnce` [access mode][k8s-accessmode].
+- MUST NOT be backed by local or ephemeral storage.
+- MUST NOT be bound to the lifecycle of a Kubernetes node.
+
+The provisioned storage class MAY support volume expansion (`allowVolumeExpansion=true`).
+
+## Previous standard versions
+
+[Version v1 of this standard](scs-0211-v1-kaas-default-storage-class.md) did not enforce the
+existence of a default storage class in a newly created cluster.
+
+Previously, the backing storage of the default storage class was required to be protected
+against data loss caused by a physical disk or host failure.
+It also contained recommendations (MAY) with regard to redundant storage across hosts
+or availability zones.
+In this revision, these requirements and recommendations have been dropped.
```

Comment:
This clarification actually confused me a bit, as I was under the impression that protection "against data loss caused by a physical disk or host failure" was intended to be covered by "MUST NOT be backed by local or ephemeral storage".

Regarding edge setups, you concluded in #652:

> - the requirement of "not being backed by local storage" should remain
> ...
> - there was a quick discussion whether this is too strict from a standards point of view
>   - example: edge clouds probably cannot fulfill this -> but this a very special case and might be solved by another hypothetical certificate scope for edge environments

There are not that many options (or use cases) to implement storage that is NOT being backed by local storage, yet NOT being protected against data loss due to disk/host failure.
So this concession to CSP's will probably create little value for CSP's, yet would reduce the value for users who may want to rely on SCS standards to be "sensible/common defaults". I do not know of any major block storage cloud offering that does not provide basic redundancy or at least snapshotting.

Reply:
> This clarification actually confused me a bit, as I was under the impression that protection "against data loss caused by a physical disk or host failure" was intended to be covered by "MUST NOT be backed by local or ephemeral storage".

I removed the "physical disk or host failure" sentence because we will not be able to test this as part of the conformance checks anyway.

My reasoning behind this is: even if you use the Cinder CSI, or let's say, the Rook CSI provisioner, the backing storage could in theory be misconfigured wrt to single host or disk failure tolerance, but there is, AFAIK, no reliable way to check this from inside the K8s cluster (and with minimal privileges).

Of course, our expectation would be that this failure tolerance is configured correctly by the CSP. As a compromise I can add a sentence that documents this expectation.

In my understanding, the important bit of this standard is that a provisioned volume isn't bound to the node's lifecycle, i.e., it doesn't just use something like the `rancher.io/local-path` or the `lvm.csi.metal-stack.io` provisioner.

For the test, I currently plan to check against an allowlist of provisioners (e.g., `cinder.csi.openstack.org`, `*.rbd.csi.ceph.com`, `driver.longhorn.io`, may need extension in the future), because I think it is the most pragmatic approach. (Going one step further, the test could even try to mount the same PVC on two different nodes.)

> So this concession to CSP's will probably create little value for CSP's, yet would reduce the value for users who may want to rely on SCS standards to be "sensible/common defaults". I do not know of any major block storage cloud offering that does not provide basic redundancy or at least snapshotting.

I do not agree completely: there is a value for the user if we prevent that a "simple" provisioner is used (like local-path), e.g., volume is independent of the node, can be used on another node, etc.

Reply:
Regarding testability: Alright, did not know that this that much of an important factor. :+1: 

---

> I do not agree completely: there is a value for the user if we prevent that a "simple" provisioner is used (like local-path), e.g., volume is independent of the node, can be used on another node, etc.

Definitely! Yet, it's kind of a niche use case.

In my experience, one would want to have either...

- some *failure-safe* volume for any kind of application. Maybe a trivial non-HA database installation for non-HA use cases. While availability may be of little concern, data durability often will be, making a storage backend with some hands-off redundancy very much favorable. This is the default on most clouds, which will also work with most use cases involving applications implementing data high availability/durability themselves.  
OR
- some *not necessarily failure-safe* volume (may be local/node-bound/non-redundant, possibly fast) to run applications that take care of data durability and availability on application level.

While offering a default storage class which is not redundant/failure-safe/replicated in any way, only non-local, would make e. g. "A lot of third-party software, such as Helm charts, [which] assume that a default storage class is configured" technically initially installable, it would probably also break assumptions of these charts regarding data durability, undermining this standard's motivation for an "out-of-the-box working experience" (at least some time later if/when a disk eventually fails).

That may be the outcome (yet I do not know of CSP's who are really restrained here), but no matter what, IMHO it should be spelled out more explicitly also in the "Decision" section.